### ISSUE TITLE
[syzkaller] KASAN: use-after-free Read in tcp_check_space

### ISSUE THREAD

**Issue by cpaasch** [multipath-tcp/mptcp_net-next#101]
```BUG: KASAN: use-after-free in instrument_atomic_read include/linux/instrumented.h:56 [inline]
BUG: KASAN: use-after-free in test_bit include/asm-generic/bitops/instrumented-non-atomic.h:110 [inline]
BUG: KASAN: use-after-free in tcp_check_space+0x5c/0x6c0 net/ipv4/tcp_input.c:5365
Read of size 8 at addr ffff88803e0dc688 by task ksoftirqd/0/9

CPU: 0 PID: 9 Comm: ksoftirqd/0 Not tainted 5.9.0-rc8 #28
Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.13.0-0-gf21b5a4aeb02-prebuilt.qemu.org 04/01/2014
Call Trace:
 __dump_stack lib/dump_stack.c:77 [inline]
 dump_stack+0xbe/0xfe lib/dump_stack.c:118
 print_address_description.constprop.0+0x3e/0x60 mm/kasan/report.c:383
 __kasan_report mm/kasan/report.c:513 [inline]
 kasan_report.cold+0x1f/0x37 mm/kasan/report.c:530
 check_memory_region_inline mm/kasan/generic.c:186 [inline]
 check_memory_region+0x14e/0x1b0 mm/kasan/generic.c:192
 instrument_atomic_read include/linux/instrumented.h:56 [inline]
 test_bit include/asm-generic/bitops/instrumented-non-atomic.h:110 [inline]
 tcp_check_space+0x5c/0x6c0 net/ipv4/tcp_input.c:5365
 tcp_data_snd_check net/ipv4/tcp_input.c:5375 [inline]
 tcp_rcv_established+0x82d/0x1c80 net/ipv4/tcp_input.c:5867
 tcp_v4_do_rcv+0x602/0x8b0 net/ipv4/tcp_ipv4.c:1652
 tcp_v4_rcv+0x252e/0x2e70 net/ipv4/tcp_ipv4.c:2034
 ip_protocol_deliver_rcu+0x2b/0x200 net/ipv4/ip_input.c:204
 ip_local_deliver_finish net/ipv4/ip_input.c:231 [inline]
 NF_HOOK include/linux/netfilter.h:409 [inline]
 ip_local_deliver+0x2e1/0x3a0 net/ipv4/ip_input.c:252
 dst_input include/net/dst.h:449 [inline]
 ip_rcv_finish net/ipv4/ip_input.c:428 [inline]
 ip_rcv_finish net/ipv4/ip_input.c:414 [inline]
 NF_HOOK include/linux/netfilter.h:409 [inline]
 ip_rcv+0xef/0x140 net/ipv4/ip_input.c:539
 __netif_receive_skb_one_core+0x197/0x1e0 net/core/dev.c:5311
 __netif_receive_skb+0x27/0x1c0 net/core/dev.c:5425
 process_backlog+0x1ec/0x6f0 net/core/dev.c:6315
 napi_poll net/core/dev.c:6759 [inline]
 net_rx_action+0x402/0xdf0 net/core/dev.c:6829
 __do_softirq+0x18c/0x644 kernel/softirq.c:298
 run_ksoftirqd kernel/softirq.c:652 [inline]
 run_ksoftirqd+0x15/0x30 kernel/softirq.c:644
 smpboot_thread_fn+0x3f5/0x860 kernel/smpboot.c:165
 kthread+0x34c/0x460 kernel/kthread.c:292
 ret_from_fork+0x22/0x30 arch/x86/entry/entry_64.S:294

Allocated by task 29123:
 kasan_save_stack+0x1b/0x40 mm/kasan/common.c:48
 kasan_set_track mm/kasan/common.c:56 [inline]
 __kasan_kmalloc mm/kasan/common.c:461 [inline]
 __kasan_kmalloc.constprop.0+0xc2/0xd0 mm/kasan/common.c:434
 slab_post_alloc_hook mm/slab.h:518 [inline]
 slab_alloc_node mm/slub.c:2888 [inline]
 slab_alloc mm/slub.c:2896 [inline]
 kmem_cache_alloc+0x156/0x490 mm/slub.c:2901
 sock_alloc_inode+0x18/0x1b0 net/socket.c:253
 alloc_inode+0x63/0x1e0 fs/inode.c:232
 new_inode_pseudo+0x14/0x120 fs/inode.c:928
 sock_alloc+0x3c/0x260 net/socket.c:573
 __sock_create+0x7d/0x440 net/socket.c:1391
 mptcp_subflow_create_socket+0xf8/0x6b0 net/mptcp/subflow.c:1177
 __mptcp_socket_create net/mptcp/protocol.c:107 [inline]
 mptcp_init_sock+0x140/0x730 net/mptcp/protocol.c:1957
 inet_create net/ipv4/af_inet.c:383 [inline]
 inet_create+0x7c7/0xbd0 net/ipv4/af_inet.c:248
 __sock_create+0x1c2/0x440 net/socket.c:1427
 sock_create net/socket.c:1478 [inline]
 __sys_socket+0xef/0x200 net/socket.c:1520
 __do_sys_socket net/socket.c:1529 [inline]
 __se_sys_socket net/socket.c:1527 [inline]
 __x64_sys_socket+0x6f/0xb0 net/socket.c:1527
 do_syscall_64+0x33/0x40 arch/x86/entry/common.c:46
 entry_SYSCALL_64_after_hwframe+0x44/0xa9

Freed by task 1251:
 kasan_save_stack+0x1b/0x40 mm/kasan/common.c:48
 kasan_set_track+0x1c/0x30 mm/kasan/common.c:56
 kasan_set_free_info+0x1b/0x30 mm/kasan/generic.c:355
 __kasan_slab_free+0x112/0x160 mm/kasan/common.c:422
 slab_free_hook mm/slub.c:1544 [inline]
 slab_free_freelist_hook mm/slub.c:1577 [inline]
 slab_free mm/slub.c:3138 [inline]
 kmem_cache_free+0x78/0x3e0 mm/slub.c:3154
 i_callback+0x42/0x70 fs/inode.c:221
 rcu_do_batch kernel/rcu/tree.c:2430 [inline]
 rcu_core+0x4e5/0x1580 kernel/rcu/tree.c:2658
 __do_softirq+0x18c/0x644 kernel/softirq.c:298

Last call_rcu():
 kasan_save_stack+0x1b/0x40 mm/kasan/common.c:48
 kasan_record_aux_stack+0xa7/0xb0 mm/kasan/generic.c:346
 __call_rcu kernel/rcu/tree.c:2896 [inline]
 call_rcu+0x76/0xac0 kernel/rcu/tree.c:2970
 destroy_inode+0x12b/0x1b0 fs/inode.c:287
 iput_final fs/inode.c:1652 [inline]
 iput fs/inode.c:1678 [inline]
 iput+0x466/0x7f0 fs/inode.c:1664
 mptcp_close+0x331/0x880 net/mptcp/protocol.c:2150
 inet_release+0xe9/0x1f0 net/ipv4/af_inet.c:434
 __sock_release+0xd2/0x280 net/socket.c:596
 sock_close+0x15/0x20 net/socket.c:1277
 __fput+0x276/0x960 fs/file_table.c:281
 task_work_run+0x109/0x1d0 kernel/task_work.c:141
 tracehook_notify_resume include/linux/tracehook.h:188 [inline]
 exit_to_user_mode_loop kernel/entry/common.c:165 [inline]
 exit_to_user_mode_prepare+0x136/0x140 kernel/entry/common.c:192
 syscall_exit_to_user_mode+0x26/0x170 kernel/entry/common.c:267
 entry_SYSCALL_64_after_hwframe+0x44/0xa9

The buggy address belongs to the object at ffff88803e0dc680
 which belongs to the cache sock_inode_cache of size 704
The buggy address is located 8 bytes inside of
 704-byte region [ffff88803e0dc680, ffff88803e0dc940)
The buggy address belongs to the page:
page:000000003d528c93 refcount:1 mapcount:0 mapping:0000000000000000 index:0x0 pfn:0x3e0dc
head:000000003d528c93 order:2 compound_mapcount:0 compound_pincount:0
flags: 0x100000000010200(slab|head)
raw: 0100000000010200 dead000000000100 dead000000000122 ffff88811aab6640
raw: 0000000000000000 0000000080130013 00000001ffffffff 0000000000000000
page dumped because: kasan: bad access detected

Memory state around the buggy address:
 ffff88803e0dc580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 ffff88803e0dc600: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc
>ffff88803e0dc680: fa fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
                      ^
 ffff88803e0dc700: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
 ffff88803e0dc780: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
==================================================================
```

Reproducer:
```
# {Threaded:false Collide:false Repeat:true RepeatTimes:0 Procs:8 Sandbox:none Fault:false FaultCall:-1 FaultNth:0 Leak:false NetInjection:false NetDevices:true NetReset:false Cgroups:false BinfmtMisc:false CloseFDs:true KCSAN:false DevlinkPCI:false USB:false VhciInjection:false Wifi:false UseTmpDir:false HandleSegv:false Repro:false Trace:false}
r0 = socket$inet_mptcp(0x2, 0x1, 0x106)
bind$inet(r0, &(0x7f00000013c0)={0x2, 0x4e20, @multicast2}, 0x10)
connect$inet(r0, &(0x7f0000000040)={0x2, 0x4e20, @loopback}, 0x4d)
sched_setscheduler(0x0, 0x2, 0x0)
sched_setaffinity(0x0, 0x0, 0x0)
setpriority(0x2, 0x0, 0xbc)
sendmsg$inet(r0, &(0x7f0000000280)={0x0, 0x0, &(0x7f0000000000)=[{&(0x7f0000000080)="ff", 0x3727d1}], 0x1}, 0x0)
r1 = socket$inet6_tcp(0xa, 0x1, 0x0)
setsockopt$inet6_group_source_req(0xffffffffffffffff, 0x29, 0x2e, 0x0, 0x0)
epoll_create(0x7)
r2 = syz_open_procfs(0x0, 0x0)
epoll_ctl$EPOLL_CTL_ADD(0xffffffffffffffff, 0x1, 0xffffffffffffffff, 0x0)
sendmmsg$unix(r2, 0x0, 0x0, 0x4)
open_by_handle_at(0xffffffffffffffff, 0x0, 0xa00)
setsockopt$inet6_tcp_TCP_FASTOPEN_KEY(r1, 0x6, 0x21, 0x0, 0x0)
recvmmsg(r0, 0x0, 0x0, 0x0, 0x0)
```

**Comment by matttbe** [comment]
Hi Christoph,

Thank you for the bug report!

By chance, do you still have the HEAD commit? A few patches have been applied end of last week.

**Comment by cpaasch** [comment]
I will give it a try.

**Comment by matttbe** [comment]
This should be fixed thanks to @pabeni 's patch: 51c3efe449ec

Please close this ticket once syzkaller can validate this bug cannot be reproduced :)

**Comment by mjmartineau** [comment]
The reproducer didn't trigger the KASAN fault after ~500000 iterations (only saw the warning from #70).
